presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java

```diff
@@ -106,153 +131,171 @@ public void doFilter(ServletRequest servletRequest, ServletResponse servletRespo
-            environment.put(SECURITY_CREDENTIALS, password);
-
-            CacheLoader<Hashtable<String, String>, DirContext> loader = new CacheLoader<Hashtable<String, String>, DirContext>()
-            {
-                @Override
-                public DirContext load(Hashtable<String, String> environment)
-                        throws NamingException
-                {
-                    return authenticate(environment);
-                }
-            };
-
-            LoadingCache authenticationCache = CacheBuilder.newBuilder().expireAfterWrite(ldapCacheTtl, MILLISECONDS).build(loader);
-            try {
-                context = (DirContext) authenticationCache.get(environment);
-            }
-            catch (ExecutionException e) {
-                Throwable throwable = e.getCause();
-                if (throwable instanceof AuthenticationException) {
-                    throw new AuthenticationException(throwable.getMessage());
-                }
-                else {
-                    throw Throwables.propagate(e);
-                }
-            }
-
-            log.debug("Authentication successful for user %s.", user);
-
-            if (genericLdapBinder.getGroupAuthorizationSearchPattern().isPresent()) {
-                checkForGroupMembership(genericLdapBinder.getGroupAuthorizationSearchPattern().get(), user, context);
-            }
-
+            String header = request.getHeader(AUTHORIZATION);
+            Credentials credentials = getCredentials(header);
+            Principal principal = getPrincipal(credentials);
             // ldap authentication ok, continue
             nextFilter.doFilter(new HttpServletRequestWrapper(request)
             {
                 @Override
                 public Principal getUserPrincipal()
                 {
-                    return new LdapPrincipal(user);
+                    return principal;
                 }
             }, servletResponse);
         }
         catch (AuthenticationException e) {
-            // log the failure since the exception goes directly to the client
+            log.debug(e, "LDAP authentication failed");
+            if (e.getStatus() == UNAUTHORIZED) {
+                sendChallenge(request, response, e.getMessage());
+                return;
+            }
+            response.sendError(e.getStatus().code(), e.getMessage());
+        }
+    }
+
+    private static void sendChallenge(HttpServletRequest request, HttpServletResponse response, String errorMessage)
+            throws IOException
+    {
+        long contentLength = request.getContentLengthLong();
+        if (contentLength > 0) {
+            skipFully(request.getInputStream(), contentLength);
+        }
+        response.setHeader(HttpHeaders.WWW_AUTHENTICATE, "Basic realm=\"presto\"");
+        response.sendError(UNAUTHORIZED.code(), errorMessage);
+    }
+
+    private static Credentials getCredentials(String header)
```

Comment:
I think those commits should be squashed anyway. I don't think that Facebook is going to merge the caches as a local field.